materialize-bigquery/schema.go

```diff
@@ -0,0 +1,292 @@
+		if err != nil {
+			return err
+		}
+		return sqlDriver.ValidateSelectedFields(constraints, t.spec)
+	}
+}
+
+// Whether this Table is configured to be only receive
+// delta updates
+func (t *Table) DeltaUpdates() bool {
+	return t.deltaUpdates
+}
+
+// The name of the Table as created in BigQuery
+func (t *Table) Name() string {
+	return t.name
+}
+
+// Field represents a field in BigQuery. The underlying Projection
+// needs to be analyzed and sanitized so that it can be safely used inside
+// BigQuery.
+type Field struct {
+	name      string
+	fieldType bigquery.FieldType
+
+	// Sanitizers are executed in the order
+	// they are added in the slice.
+	sanitizers []sanitizerFunc
+}
+
+type sanitizerFunc func(string) string
+
+// NewField returns a field constructed from a projection provided by a BindingSpec
+// The Field returned includes logic that allows a field to map a Key and its associated value
+// into SQL safely. The field gets sanitized and quoted according to the type of field
+// bigquery will see.
+func NewField(projection *pf.Projection) (*Field, error) {
+	f := &Field{
+		name:       identifierSanitizer(projection.Field),
+		sanitizers: []sanitizerFunc{},
+	}
+
+	// This is a test to make sure that there is only 1 infered type for this projection
+	// ie. Inference.Types = [string, integer] is wrong. [string, null] is OK. [string, integer, null] is wrong
+	if !projection.Inference.IsSingleType() {
+		return nil, fmt.Errorf("don't expect multiple types besides Null and a real type, can't proceed: %s", &projection.Inference.Types)
+	}
+
+	for _, p := range projection.Inference.Types {
+		switch p {
+		// The string format can be further expanded so that it can be stored in a more specific format
+		// When a projection is a string and it has the Inference_String set, it's possible for the user
+		// to pass along more information about the string field to let the connector create a more
+		// appropriate field for that string. If an Inference_String{} is present, it doesn't default back
+		// to a normal string as it is prefered to let the user know that the inference could not be matched.
+		case "string":
+			if projection.Inference.String_ != nil {
+				s := projection.Inference.String_
+				switch s.Format {
+				case "date-time":
+					f.fieldType = bigquery.TimestampFieldType
+				case "date":
+					f.fieldType = bigquery.DateFieldType
+				// There is a case where a String_ is present but the values aren't set.
+				// I'm not sure how this can happen since the protobuf struct should
+				// be optional and nil, but it did happen in my testing and I couldn't
+				// figured out why, so this is here as a precaution (P-O)
+				case "":
+					f.fieldType = bigquery.StringFieldType
+					f.sanitizers = append(f.sanitizers, sqlDriver.DefaultQuoteSanitizer)
+
+				default:
+					return nil, fmt.Errorf("inferring string column type as the format is not supported: %s", s.Format)
```

Comment:
I think we should treat the previous case (no format) as the default, and not return an error here. It's expected that there will be format identifiers that we won't be able to recognize.